# Проект видеоигр

- Автор: Козбаков Айдын
- Дата: 11.02.2026г.

### Цели и задачи проекта

познакомиться с данными, проверить их корректность и провести предобработку, получив необходимый срез данных

### Описание данных

Данные /datasets/new_games.csv содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:
- Name — название игры.
- Platform — название платформы.
- Year of Release — год выпуска игры.
- Genre — жанр игры.
- NA sales — продажи в Северной Америке (в миллионах проданных копий).
- EU sales — продажи в Европе (в миллионах проданных копий).
- JP sales — продажи в Японии (в миллионах проданных копий).
- Other sales — продажи в других странах (в миллионах проданных копий).
- Critic Score — оценка критиков (от 0 до 100).
- User Score — оценка пользователей (от 0 до 10).
- Rating — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

## 1. Загрузка данных и знакомство с ними

In [1]:
import pandas as pd

try:
    df = pd.read_csv('new_games.csv') 
except FileNotFoundError:
    df = pd.read_csv('/datasets/games.csv')

In [2]:
df.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


Датасет `new_games.csv` содержит 11 столбцов и 16956 строк, в которых представлена информация о продажах видеоигр.

Изучим данные и их корректность

- **Строковые значения (str)**- 7 столбцов: Name, Platform, Genre, EU sales, JP sales, User score, Rating. Здесь тип данных не подходит для кол-ва продаж, оценки от пользователей и рейтинга
- **Числовые значения с плавающей запятой (float64)** - 4 столбца: Year of Release, NA sales, Other sales, Critic Score. Тип данных так же не подходит для года выпуска, кол-ва продаж.

После анализа данных видно, что большинство столбцов представлены не корректно. Для оптимизации требуется преобразовать строковые данные и числовые с плавающей запятой в целочисленные

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма


In [4]:
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [5]:
# Переводим названия столбцов в стиль snake case
df.columns = df.columns.str.lower().str.replace(' ','_')

In [6]:
df.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

### 2.2. Типы данных


In [7]:
# переведем год выпуска в числовой формат int64 т.к. у нас есть только год, без конкретной даты
df['year_of_release'] = df['year_of_release'].astype('Int64')
df['year_of_release']

0        2006
1        1985
2        2008
3        2009
4        1996
         ... 
16951    2016
16952    2006
16953    2016
16954    2003
16955    2016
Name: year_of_release, Length: 16956, dtype: Int64

In [8]:
# переведем значения в полях с продажами и оценкой пользователей в float
df[['eu_sales', 'jp_sales','user_score']] = df[['eu_sales', 'jp_sales','user_score']].apply(pd.to_numeric, errors = 'coerce')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  Int64  
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16950 non-null  float64
 6   jp_sales         16952 non-null  float64
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       7688 non-null   float64
 10  rating           10085 non-null  object 
dtypes: Int64(1), float64(6), object(4)
memory usage: 1.4+ MB


мы перевели значения в float, заменив значения unkown на NaN

- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

### 2.3. Наличие пропусков в данных


In [9]:
# кол-во пропусков
df.isna().sum()

name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64

In [10]:
## относительное кол-во пропусков
df.isna().sum() / len(df) * 100

name                0.011795
platform            0.000000
year_of_release     1.621845
genre               0.011795
na_sales            0.000000
eu_sales            0.035386
jp_sales            0.023590
other_sales         0.000000
critic_score       51.391838
user_score         54.659118
rating             40.522529
dtype: float64

- Пропуски встречаются в name, year of release, genre, critic score, user score, rating
- Больше всего пропуски в critic score (8714 - 51,4%), user score и rating (около 6800 - 40%) что является значительным показателем
- Данные в этих столбцах могут отсутствовать из-за того что ESRB рейтинг был принят в 1994 году, и некоторые игры были выпущены до этого и не получили свой рейтинг, то же самое может быть и с оценками
- можно удалить строки где есть пропуски у столбцов name, year of release, genre так как число пропусков не превышает 2%
- для столбцов с оценками и рейтинга пропуски можно заменить на индикатор или "нет оценки", мы не можем удалить эти значения (они составляют половину данных)

In [11]:
# Удаляем строки, где пропущены данные
df = df.dropna(subset=['name', 'genre', 'year_of_release'])

In [12]:
df.isna().sum()

name                  0
platform              0
year_of_release       0
genre                 0
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8594
user_score         9121
rating             6778
dtype: int64

In [14]:
# Заполняем пропуски в продажах средним значением для той же платформы и года
for col in ['eu_sales', 'jp_sales']:
    df[col] = df[col].fillna(df.groupby(['platform', 'year_of_release'])[col].transform('mean'))
    df[col] = df[col].fillna(0)
    # Если после этого остались пропуски (например, игра единственная на платформе за этот год), 
# просто заполняем их нулями




In [15]:
df.isna().sum()

name                  0
platform              0
year_of_release       0
genre                 0
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score       8594
user_score         9121
rating             6778
dtype: int64

In [16]:
# заменим пропуски в рейтинге на unknown
df['rating'] = df['rating'].fillna('unknown')

In [17]:
df[['user_score','critic_score']] = df[['user_score','critic_score']].fillna(-1)

### 2.4. Явные и неявные дубликаты в данных

In [18]:
# сделаем названия игр с заглавной буквы
df['name']= df['name'].str.title()

In [19]:
df.groupby('name')['name'].count()

name
 Beyblade Burst                                  2
 Fire Emblem Fates                               1
 Frozen: Olaf'S Quest                            2
 Haikyu!! Cross Team Match!                      1
 Tales Of Xillia 2                               1
                                                ..
Zwei!!                                           1
Zyuden Sentai Kyoryuger: Game De Gaburincho!!    1
[Prototype 2]                                    3
[Prototype]                                      2
¡Shin Chan Flipa En Colores!                     1
Name: name, Length: 11426, dtype: int64

In [20]:
# находим количество жанров
df.groupby('genre')['genre'].count()

genre
ACTION            12
ADVENTURE          4
Action          3341
Adventure       1309
FIGHTING           6
Fighting         844
MISC               3
Misc            1740
PLATFORM           3
PUZZLE             2
Platform         891
Puzzle           577
RACING             6
ROLE-PLAYING       6
Racing          1244
Role-Playing    1493
SHOOTER            5
SIMULATION         1
SPORTS             7
STRATEGY           1
Shooter         1314
Simulation       866
Sports          2325
Strategy         679
Name: genre, dtype: int64

In [21]:
# сделаем жанры с заглавной буквы
df['genre']= df['genre'].str.title()

In [22]:
df.groupby('genre')['genre'].count()

genre
Action          3353
Adventure       1313
Fighting         850
Misc            1743
Platform         894
Puzzle           579
Racing          1250
Role-Playing    1499
Shooter         1319
Simulation       867
Sports          2332
Strategy         680
Name: genre, dtype: int64

In [23]:
for column in ['platform','year_of_release','genre','rating']:
    print(f'Уникальные значения в столбце {column}:')
    print(df[column].sort_values().unique())
    print()

Уникальные значения в столбце platform:
['2600' '3DO' '3DS' 'DC' 'DS' 'GB' 'GBA' 'GC' 'GEN' 'GG' 'N64' 'NES' 'NG'
 'PC' 'PCFX' 'PS' 'PS2' 'PS3' 'PS4' 'PSP' 'PSV' 'SAT' 'SCD' 'SNES' 'TG16'
 'WS' 'Wii' 'WiiU' 'X360' 'XB' 'XOne']

Уникальные значения в столбце year_of_release:
<IntegerArray>
[1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992,
 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005,
 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016]
Length: 37, dtype: Int64

Уникальные значения в столбце genre:
['Action' 'Adventure' 'Fighting' 'Misc' 'Platform' 'Puzzle' 'Racing'
 'Role-Playing' 'Shooter' 'Simulation' 'Sports' 'Strategy']

Уникальные значения в столбце rating:
['AO' 'E' 'E10+' 'EC' 'K-A' 'M' 'RP' 'T' 'unknown']



In [24]:
# находим количество платформ
df.groupby('platform')['platform'].count()

platform
2600     118
3DO        3
3DS      522
DC        52
DS      2147
GB        97
GBA      826
GC       549
GEN       27
GG         1
N64      320
NES      100
NG        12
PC       972
PCFX       1
PS      1208
PS2     2154
PS3     1330
PS4      395
PSP     1212
PSV      434
SAT      174
SCD        6
SNES     241
TG16       2
WS         6
Wii     1305
WiiU     147
X360    1249
XB       818
XOne     251
Name: platform, dtype: int64

In [25]:
# ищем количество рейтингов
df.groupby('rating')['rating'].count()

rating
AO            1
E          3968
E10+       1414
EC            8
K-A           3
M          1558
RP            1
T          2948
unknown    6778
Name: rating, dtype: int64

In [26]:
# заменим рейтинг К-А на Е, так как К-А устарел
df['rating'] = df['rating'].str.replace('K-A', 'E')

In [27]:
# ищем количество игр по годам
df.groupby('year_of_release')['year_of_release'].count()

year_of_release
1980       9
1981      46
1982      37
1983      18
1984      14
1985      14
1986      22
1987      17
1988      15
1989      17
1990      16
1991      42
1992      43
1993      60
1994     121
1995     220
1996     267
1997     293
1998     384
1999     341
2000     357
2001     491
2002     839
2003     789
2004     771
2005     950
2006    1020
2007    1218
2008    1445
2009    1450
2010    1279
2011    1149
2012     670
2013     552
2014     584
2015     612
2016     507
Name: year_of_release, dtype: int64

In [28]:
# выявляем количество явных дубликатов в датафрейме
df.duplicated().sum()

235

In [29]:
# количество строк ДО удаления
rows_before = len(df)

# удаляем дубликаты и сбрасываем индексы
df.drop_duplicates(inplace=True)
df = df.reset_index(drop=True)

# сохраняем количество строк ПОСЛЕ удаления
rows_after = len(df)

# считаем разницу
absolute_deleted = rows_before - rows_after
relative_deleted = (absolute_deleted / rows_before) * 100

# выводим результат
print(f"Удалено явных дубликатов (абсолютное значение): {absolute_deleted}")
print(f"Процент удаленных дубликатов (относительное значение): {relative_deleted:.2f}%")

Удалено явных дубликатов (абсолютное значение): 235
Процент удаленных дубликатов (относительное значение): 1.41%


In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16444 entries, 0 to 16443
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16444 non-null  object 
 1   platform         16444 non-null  object 
 2   year_of_release  16444 non-null  Int64  
 3   genre            16444 non-null  object 
 4   na_sales         16444 non-null  float64
 5   eu_sales         16444 non-null  float64
 6   jp_sales         16444 non-null  float64
 7   other_sales      16444 non-null  float64
 8   critic_score     16444 non-null  float64
 9   user_score       16444 non-null  float64
 10  rating           16444 non-null  object 
dtypes: Int64(1), float64(6), object(4)
memory usage: 1.4+ MB


- После проведения предобработки данных напишите общий промежуточный вывод.

- Изначально в датафрейме было 16956 записей
- Сейчас записей стало 16444
- было удалено 512 строк
- из них 235 явных дупликатов

---

## 3. Фильтрация данных


In [31]:
df_actual = df.query('2000 <= year_of_release <= 2013').reset_index(drop=True)

In [32]:
df_actual.head()

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,Wii Sports,Wii,2006,Sports,41.36,28.96,3.77,8.45,76.0,8.0,E
1,Mario Kart Wii,Wii,2008,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
2,Wii Sports Resort,Wii,2009,Sports,15.61,10.93,3.28,2.95,80.0,8.0,E
3,New Super Mario Bros.,DS,2006,Platform,11.28,9.14,6.50,2.88,89.0,8.5,E
4,Wii Play,Wii,2006,Misc,13.96,9.18,2.93,2.84,58.0,6.6,E


---

## 4. Категоризация данных


In [33]:
df_actual['user_score_category'] = pd.cut(df_actual['user_score'], bins=[-1,0,3,8,10], labels=['Без оценки','Низкая оценка','Средняя оценка','Высокая оценка'], right = False)

In [34]:
df_actual.head(20)

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating,user_score_category
0,Wii Sports,Wii,2006,Sports,41.36,28.96,3.77,8.45,76.0,8.0,E,Высокая оценка
1,Mario Kart Wii,Wii,2008,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E,Высокая оценка
2,Wii Sports Resort,Wii,2009,Sports,15.61,10.93,3.28,2.95,80.0,8.0,E,Высокая оценка
3,New Super Mario Bros.,DS,2006,Platform,11.28,9.14,6.50,2.88,89.0,8.5,E,Высокая оценка
4,Wii Play,Wii,2006,Misc,13.96,9.18,2.93,2.84,58.0,6.6,E,Средняя оценка
5,New Super Mario Bros. Wii,Wii,2009,Platform,14.44,6.94,4.70,2.24,87.0,8.4,E,Высокая оценка
6,Nintendogs,DS,2005,Simulation,9.05,10.95,1.93,2.74,-1.0,-1.0,unknown,Без оценки
7,Mario Kart Ds,DS,2005,Racing,9.71,7.47,4.13,1.90,91.0,8.6,E,Высокая оценка
8,Wii Fit,Wii,2007,Sports,8.92,8.03,3.60,2.15,80.0,7.7,E,Средняя оценка
9,Kinect Adventures!,X360,2010,Misc,15.00,4.89,0.24,1.69,61.0,6.3,E,Средняя оценка


In [35]:
df_actual['critic_score_category'] = pd.cut(df_actual['critic_score'], bins=[-1,0,30,80,100], labels=['Без оценки','Низкая оценка','Средняя оценка','Высокая оценка'], right = False)

In [36]:
df_actual.groupby('user_score_category')['user_score_category'].count()

user_score_category
Без оценки        6298
Низкая оценка      116
Средняя оценка    4081
Высокая оценка    2286
Name: user_score_category, dtype: int64

In [37]:
df_actual.groupby('critic_score_category')['critic_score_category'].count()

critic_score_category
Без оценки        5612
Низкая оценка       55
Средняя оценка    5422
Высокая оценка    1692
Name: critic_score_category, dtype: int64

In [38]:
df_actual.head(10)

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating,user_score_category,critic_score_category
0,Wii Sports,Wii,2006,Sports,41.36,28.96,3.77,8.45,76.0,8.0,E,Высокая оценка,Средняя оценка
1,Mario Kart Wii,Wii,2008,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E,Высокая оценка,Высокая оценка
2,Wii Sports Resort,Wii,2009,Sports,15.61,10.93,3.28,2.95,80.0,8.0,E,Высокая оценка,Высокая оценка
3,New Super Mario Bros.,DS,2006,Platform,11.28,9.14,6.50,2.88,89.0,8.5,E,Высокая оценка,Высокая оценка
4,Wii Play,Wii,2006,Misc,13.96,9.18,2.93,2.84,58.0,6.6,E,Средняя оценка,Средняя оценка
5,New Super Mario Bros. Wii,Wii,2009,Platform,14.44,6.94,4.70,2.24,87.0,8.4,E,Высокая оценка,Высокая оценка
6,Nintendogs,DS,2005,Simulation,9.05,10.95,1.93,2.74,-1.0,-1.0,unknown,Без оценки,Без оценки
7,Mario Kart Ds,DS,2005,Racing,9.71,7.47,4.13,1.90,91.0,8.6,E,Высокая оценка,Высокая оценка
8,Wii Fit,Wii,2007,Sports,8.92,8.03,3.60,2.15,80.0,7.7,E,Средняя оценка,Высокая оценка
9,Kinect Adventures!,X360,2010,Misc,15.00,4.89,0.24,1.69,61.0,6.3,E,Средняя оценка,Средняя оценка


In [39]:
df_actual.groupby('platform')['platform'].count().sort_values(ascending=False).head(7)

platform
PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1087
GBA      811
Name: platform, dtype: int64

---

## 5. Итоговый вывод


- Мы можем сделать вывод, что за период 2000 - 2013 год самая популярная платформа по кол-ву выпущенных игр является **ps2** (2127) и **DS** (2120), наименее популярная GBA (811)
- Мы проделали большую работу по обработке данных: преобразовали типы данных, привели названия к нужному стилю, исправили пропуски, дубликаты, сделали категоризацию по оценкам и годам, и финально узнали топ 7 платформ по выпуску игр
- Так же добавили два столбца с категорией оценки от пользователей и критиков